overlord/devicestate/handlers_systems.go

```diff
@@ -121,6 +132,121 @@ func purgeNewSystemSnapFiles(logfile string) error {
 	return s.Err()
 }
 
+func snapsUniqueToRecoverySystem(target string, systems []*System) ([]*seed.Snap, error) {
+	requiredByOtherSystems := make(map[string]bool)
+	for _, sys := range systems {
+		if sys.Label == target {
+			continue
+		}
+
+		sd, err := seed.Open(dirs.SnapSeedDir, sys.Label)
+		if err != nil {
+			return nil, err
+		}
+
+		if err := sd.LoadAssertions(nil, func(*asserts.Batch) error {
+			return nil
+		}); err != nil {
+			return nil, err
+		}
+
+		if err := sd.LoadMeta(seed.AllModes, nil, timings.New(nil)); err != nil {
+			return nil, err
+		}
+
+		err = sd.Iter(func(sn *seed.Snap) error {
+			requiredByOtherSystems[sn.Path] = true
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	targetSeed, err := seed.Open(dirs.SnapSeedDir, target)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := targetSeed.LoadAssertions(nil, func(*asserts.Batch) error {
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	if err := targetSeed.LoadMeta(seed.AllModes, nil, timings.New(nil)); err != nil {
+		return nil, err
+	}
+
+	var uniqueToTarget []*seed.Snap
+	err = targetSeed.Iter(func(sn *seed.Snap) error {
+		if !requiredByOtherSystems[sn.Path] {
+			uniqueToTarget = append(uniqueToTarget, sn)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return uniqueToTarget, nil
+}
+
+func (m *DeviceManager) doRemoveRecoverySystem(t *state.Task, _ *tomb.Tomb) error {
+	systems, err := m.Systems()
+	if err != nil {
+		return fmt.Errorf("cannot get recovery systems: %w", err)
+	}
+
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	setup, err := taskRemoveRecoverySystemSetup(t)
+	if err != nil {
+		return err
+	}
+
+	for _, sys := range systems {
+		if sys.Label == setup.Label && sys.Current {
+			return fmt.Errorf("cannot remove current recovery system: %q", setup.Label)
+		}
+	}
+
+	if len(systems) == 1 {
+		return fmt.Errorf("cannot remove last recovery system: %q", setup.Label)
+	}
+
+	deviceCtx, err := DeviceCtx(st, t, nil)
+	if err != nil {
+		return fmt.Errorf("cannot get device context: %w", err)
+	}
+
+	snapsToRemove, err := snapsUniqueToRecoverySystem(setup.Label, systems)
```

Comment:
I wonder if this handler is idempotent, since os.RemoveAll() called a bit later will recurse inside the directory and remove pieces, meaning that attempts to load the seed inside this call may fail at this point. Perhaps this information should be stored in the task and acted upon if we reenter the handler. For a scenario when creating a recovery system it is ok to fail and clean up partial tree, but here the intention is to have the data be removed.

Reply:
That makes sense, it isn't idempotent right now. Question though, when adding something to the task, how will we end up running the task handler again with the same task? I ever return anything like a `state.Retry` from the handler.

Reply:
I always that a system (especially UC) may reboot unexpectedly. Then unless the task's state is changed to Done (which is either explicit when calling task.SetStatus(), or implicitly in taskrunner), when snapd starts up it will execute the handler again.

Reply:
Makes sense! ef481329c45d016a45ab7d2a1e49dabf27f5591c does the requested changes to enable the handler to re-run.